src/qibo/transpiler/unitary_decompositions.py

```diff
@@ -64,18 +64,14 @@ def calculate_psi(unitary, backend, magic_basis=magic_basis):
     if backend.__class__.__name__ not in ("PyTorchBackend", "TensorflowBackend"):
         # eig seems to have a different behavior based on backend/hardware,
         # use np.round to increase precision seems to fix the issue
-        eigvals, psi_magic_complex = backend.calculate_eigenvectors(
-            np.round(ut_u, decimals=20), hermitian=False
-        )
         eigvals_real, psi_magic = backend.calculate_eigenvectors(
             np.round(backend.np.real(ut_u), decimals=20), hermitian=True
         )
     else:
-        eigvals, psi_magic_complex = backend.calculate_eigenvectors(ut_u, hermitian=False)
         eigvals_real, psi_magic = backend.calculate_eigenvectors(backend.np.real(ut_u), hermitian=True)
 
     # sort eigvals to match psi_magic, as eigvals_real only give real part (at least in nnumpy backend)
-    eigvals = backend.np.sort(eigvals)
+    eigvals = backend.np.sort(eigvals_real)
```

Comment:
We need to have the full complex eigenvalue. But it looks better without the extra calculate_eigenvectors, so maybe we can just compute eigvals as:
$eigvals[k] = \langle \psi_k|U^T U |\psi_k\rangle$

```suggestion
    eigvals = backend.np.sum(backend.np.multiply(backend.np.conj(psi_magic), backend.np.matmul(ut_u, psi_magic)) , 0)
```